### PR TITLE
identity improvements:

### PR DESCRIPTION
**Pull request by bryanchriswhite**
+ identity refactor
+ add `identity revocations`
+ add `identity certificate-authority revoke-peer`

```
$ identity --help
Identity management

Usage:
  /home/user/go/bin/identity [command]

Available Commands:
  revocations           Print revocation information from a revocation database
...
```

`revocations`
---
```
$ identity revocations --help
Print revocation information from a revocation database

Usage:
  /home/user/go/bin/identity revocations [service] [flags]

Flags:
  -h, --help                      help for revocations
      --revocation-dburl string   url for revocation database (e.g. bolt://some.db OR redis://127.0.0.1:6378?db=2&password=abc123) (default "bolt:///home/user/.local/share/storj/revocations.db")

Global Flags:
      --config-dir string           service config directory (default "/home/user/.local/share/storj")

```

`certificate-authority revoke-peer`
---
```
$ identity certificate-authority revoke-peer --help
Revoke a peer identity's CA certificate and add to local revocation database

Usage:
  /home/user/go/bin/identity certificate-authority revoke-peer [service] [revoked cert path] [flags]

Flags:
      --ca.cert-path string        path to the certificate chain for this identity (default "/home/user/.local/share/storj/identity/ca.cert")
      --ca.key-path string         path to the private key for this identity (default "/home/user/.local/share/storj/identity/ca.key")
  -h, --help                       help for revoke-peer
      --peer-ca.cert-path string   path to the certificate chain for this identity (default "/home/user/.local/share/storj/identity/ca.cert")
      --revocation-dburl string

Global Flags:
      --config-dir string           service config directory (default "/home/user/.local/share/storj")
```

### CODE REVIEW

File: cmd/certificates/auth.go [hunk 1]
```diff
@@ -49,6 +50,23 @@ var (
 )
 
 func init() {
+	// NB: init functions are executed in lexicographical order of filename
+	confDirParam := cfgstruct.FindConfigDirParam()
```
Comment (egonelbre) [inline cmd/certificates/auth.go]: Can we do all of the setup in main instead, that way it's clearer for order.

Comment (bryanchriswhite) [reply cmd/certificates/auth.go]: this stuff has to be here otherwise it won't be defined in time for other `init` functions

Comment (egonelbre) [reply cmd/certificates/auth.go]: I mean if we remove all init functions and do everything in `func main()` then there shouldn't be problems.

Comment (bryanchriswhite) [reply cmd/certificates/auth.go]: ah - misunderstood... sure!

File: cmd/certificates/sign.go [hunk 2]
```diff
@@ -5,59 +5,57 @@ package main
...
-	err = signCfg.CA.SaveBackup(ca)
+	err = signCfg.SigneeCACfg.SaveBackup(ca)
 	if err != nil {
 		return err
 	}
-	err = signCfg.Identity.SaveBackup(ident)
+	err = signCfg.SigneeIdentCfg.SaveBackup(ident)
 	if err != nil {
-		return err
+		log.Printf("unable to save backup of identity: %s\n", err.Error())
```
Comment (mobyvb) [inline cmd/certificates/sign.go]: why is the error on line 54 returned but the error here is just logged out? Don't we want the same behavior for each `SaveBackup` error?

File: cmd/certificates/sign.go [hunk 3]
```diff
@@ -5,59 +5,57 @@ package main
...
+	ca, err := signCfg.SigneeCACfg.Load()
 	if err != nil {
 		return err
 	}
 
-	ident, err := signCfg.Identity.Load()
+	ident, err := signCfg.SigneeIdentCfg.Load()
 	if err != nil {
-		return err
+		log.Printf("unable to load identity: %s\n", err.Error())
```
Comment (mobyvb) [inline cmd/certificates/sign.go]: similar question - why are we logging the error here and not on line 36?

File: cmd/certificates/sign.go (hunk 2 above)
Comment (bryanchriswhite) [reply cmd/certificates/sign.go]: the thought was to allow for the signing of a ca without an identity having to be present. You could always generated an identity after signing.the CA.

File: cmd/certificates/sign.go (hunk 3 above)
Comment (bryanchriswhite) [reply cmd/certificates/sign.go]: (same)

File: cmd/certificates/sign.go [hunk 4]
```diff
@@ -5,59 +5,57 @@ package main
...
 	}
 
 	signer, err := signCfg.Signer.Load()
 	if err != nil {
+		fmt.Println("two")
```
[the remark this answers is not in the document]
Comment (bryanchriswhite) [reply cmd/certificates/sign.go]: :+1:

File: pkg/identity/extensions_test.go [hunk 5]
```diff
@@ -0,0 +1,261 @@
...
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/mock"
+	"github.com/zeebo/errs"
+
+	"storj.io/storj/internal/testpeertls"
+	"storj.io/storj/pkg/peertls"
+)
+
+type extensionHandlerMock struct {
+	mock.Mock
```
Comment (egonelbre) [inline pkg/identity/extensions_test.go]: try using a field count and return nil instead of using `mock.Mock`, the code should end up nicer

File: cmd/certificates/sign.go [hunk 6]
```diff
@@ -5,43 +5,39 @@ package main
 
 import (
 	"crypto/x509"
+	"log"
```
Comment (egonelbre) [inline cmd/certificates/sign.go]: Use zap.Logger instead, mixing multiple logging approaches usually ends up looking bad on the console.

File: cmd/certificates/sign.go (hunk 2 above)
Comment (egonelbre) [reply cmd/certificates/sign.go]: If we don't have a SigneeIdentCfg specified can we avoid logging at all? The message might look confusing.

File: pkg/certificates/config.go [hunk 7]
```diff
@@ -85,7 +85,7 @@ func (c CertServerConfig) NewAuthDB() (*AuthorizationDB, error) {
 }
 
 // Run implements the responsibility interface, starting a certificate signing server.
-func (c CertServerConfig) Run(ctx context.Context, server *server.Server) (err error) {
+func (c CertServerConfig) Run(ctx context.Context, srv *server.Server) (err error) {
```
Comment (egonelbre) [inline pkg/certificates/config.go]: Let's avoid cleanups in other packages that you are not directly touching.

File: pkg/identity/extensions_test.go [hunk 8]
```diff
@@ -0,0 +1,261 @@
...
+			revokedLeafChain,
+			nil,
+			nil,
+			nil,
+		},
+		{
+			"certificate revocation - serial revocations",
+			peertls.TLSExtConfig{Revocation: true},
+			1,
+			func() []*x509.Certificate {
```
Comment (egonelbre) [inline pkg/identity/extensions_test.go]: Using func closures inside tests usually indicates that it should be a separate test. These make hard to follow what's actually happening in the test and what's being tested and in which order.

Comment (egonelbre) [reply pkg/identity/extensions_test.go]: Also asserts here will end up in the wrong subtest.

Comment (egonelbre) [reply pkg/identity/extensions_test.go]: https://github.com/storj/storj/wiki/Testing#table-tests

File: pkg/identity/extensions_test.go [hunk 9]
```diff
@@ -0,0 +1,261 @@
...
+	whitelistSignedKeys, whitelistSignedChain, err := testpeertls.NewCertChain(3)
+	assert.NoError(t, err)
+
+	err = peertls.AddSignedCertExt(whitelistSignedKeys[0], whitelistSignedChain[0])
+	assert.NoError(t, err)
+
+	_, unrelatedChain, err := testpeertls.NewCertChain(1)
+	assert.NoError(t, err)
+
+	tmp, err := ioutil.TempDir("", "TestParseExtensions")
```
[the remark this answers is not in the document]
Comment (egonelbre) [reply pkg/identity/extensions_test.go]: Ah, didn't notice that this was a move. If it's a move then we can merge this one and do the fixes in a separate PR. That way changes are easier to see.

File: pkg/utils/utils.go [hunk 10]
```diff
@@ -67,28 +69,18 @@ func (errs combinedError) Error() string {
 }
 
 // ErrorGroup contains a set of non-nil errors
-type ErrorGroup []error
+type ErrorGroup errs.Group
```
[the remark this answers is not in the document]
Comment (bryanchriswhite) [reply pkg/utils/utils.go]: Makes sense.. I'll avoid this in the future. The reason I did this was because I encountered some old code that was using `ErrGroup` and thought it would be nice to make it consistent everywhere that was still using it.

[the remark this answers is not in the document]
Comment (egonelbre) [reply pkg/utils/utils.go]: I agree, but the fix for that would be to start replacing `utils.ErrorGroup` package by package and finally remove `utils.ErrorGroup` as the last step.

File: cmd/certificates/sign.go (hunk 2 above)
Comment (bryanchriswhite) [reply cmd/certificates/sign.go]: :+1:

File: pkg/utils/utils.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (bryanchriswhite) [reply pkg/utils/utils.go]: yea, I sort of saw this as an intermediary step to keep the implementations the same.

[the remark this answers is not in the document]
Comment (bryanchriswhite) [reply pkg/utils/utils.go]: should I roll it back or is it fine to keep it?

[the remark this answers is not in the document]
Comment (egonelbre) [reply pkg/utils/utils.go]: If it doesn't cause test failures, then it's fine.

File: pkg/certificates/config.go (hunk 7 above)
Comment (bryanchriswhite) [reply pkg/certificates/config.go]: going forward I'll keep an eye out to avoid that :+1:

File: pkg/utils/utils.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (bryanchriswhite) [reply pkg/utils/utils.go]: :+1:
